### PR TITLE
post action Promise functions

### PR DESCRIPTION
**Pull request by AprilSylph**
alternate title: "One-Click Postage but the variable names make sense"

This introduces 5 new Promise functions (`fetch`, `update`, `delete`, `like` and `unlike`) to `XKit.interface`, all under a single object, `XKit.interface.post`. That was already a function, but it's a misleading name for what it does - it doesn't post anything, it parses an existing post. Thus, it is now `XKit.interface.parse_post`.

The only motivating reason for this is that constructing XHRs manually in extensions is silly when the way we interact with posts is going to be the same for all extensions at any given time. If we have and use functions to fetch, update, delete, like and unlike posts, we can update those and fix however many extensions get broken by a Tumblr update without even needing to know the full extent of what's been affected.

The most obvious starting place to me to implement these new functions is One-Click Postage, since the current code is a mess - request variables are being declared as empty objects, filled with mostly unnecessary data, then being passed to a fresh (Nx_)XHR even though we know that we're always going to need to stringify the JSON, that it's always going to be `json: true`, that it's always going to be sent via `POST`, that we always need a kitty for `/svc/post/update`, and anything else I don't need to think about if I'm looking at *just* OCP's code.
This is of course true for a lot of other extensions but OCP being the most-used extension and a swamp code-wise doesn't sit right with me at all, so I'm PRing this as it is instead of looking to clean up everything in one go.

### CODE REVIEW

**Comment by nightpool** [comment]
So all of these extensions are going to stop working for people currently on 7.8.2? Is that intentional?

**Comment by AprilSylph** [comment]
...what? I've been using 7.8.2 and it hasn't broken anything. What breaks for you?

**Comment by nightpool** [comment]
Maybe i'm confused about how these 7.9.0 patches functions are supposed to work. I tried reading the code, but object iteration order is unspecified in javascript, so I'm not sure which way it's supposed to be iterating, from low to high or high to low (do patches for 7.9.0 get run on 7.8.2?)

**Comment by AprilSylph** [comment]
it runs low to high, starting at the current version; 7.8.2 will run 7.8.2 and 7.9.0 patches but not 7.8.1, since the changes in 7.8.1 were also built directly into xkit.js

**Comment by nightpool** [comment]
well, like I said, object iteration in javascript is undefined, so it's possible that this behavior may change at any time. (or vary across browsers)

File: Extensions/one_click_postage.js [hunk 1]
```diff
@@ -974,7 +833,6 @@ XKit.extensions.one_click_postage = new Object({
 						XKit.extensions.one_click_postage.user_on_box = true;
 						XKit.extensions.one_click_postage.open_menu($(this), false, true);
 						$('#x1cpostage_tags').focus();
-						break;
```
Comment (hobinjk) [inline Extensions/one_click_postage.js]: I like keeping the `break` here so that it's less disruptive to add more cases in the future

**Comment by AprilSylph** [comment]
agreed on the `break` you singled out, but do you also mean at the end of the other switch blocks? the `default` cases on those are catch-alls so future cases won't be going under them

File: Extensions/one_click_postage.js [hunk 2]
```diff
@@ -214,237 +214,122 @@ XKit.extensions.one_click_postage = new Object({
...
-				return;
-			}
-		}
-
-		var m_object = {
-			channel_id: blog_id,
-			reblog_id: post_id,
-			reblog_key: reblog_key,
-			form_key: form_key,
-			post_type: ""
```
Comment (nightpool) [inline Extensions/one_click_postage.js]: did we remove this on purpose? XKit.interface.fetch_post seems to indicate it's required

File: Extensions/one_click_postage.js [hunk 3]
```diff
@@ -568,130 +455,105 @@ XKit.extensions.one_click_postage = new Object({
...
 
-		var m_remove_button = "<div id=\"x1cpostage_remove_caption\">remove caption</div>";
+		this.default_blog_id = this.preferences.default_blog.value || blogs[0];
+		this.blogs_list = blogs;
 
-		if (!this.preferences.show_caption_remover.value) {
-			m_remove_button = "";
+		var blog_html = "";
+		for (var x in blogs) {
+			blog_html += '<option value="' + blogs[x] + '"' + (blogs[x] == this.default_blog_id ? ' selected' : '') + '>' + blogs[x] + '</option>';
```
Comment (nightpool) [inline Extensions/one_click_postage.js]: template strings

File: Extensions/one_click_postage.js [hunk 4]
```diff
@@ -568,130 +455,105 @@ XKit.extensions.one_click_postage = new Object({
...
-		var m_remove_box_style = "";
-		if (!this.preferences.show_caption.value) {
-			m_remove_box_style = " style=\"display: none;\" ";
-		}
+		var ui_html = [
+			'<div id="x1cpostage_reblog"><i>&nbsp;</i></div>' + '<div id="x1cpostage_queue"><i>&nbsp;</i></div>' + '<div id="x1cpostage_draft"><i>&nbsp;</i></div>',
 
-		var m_clear_tags_button = "<div id=\"x1cpostage_clear_tags\">clear tags</div>";
+			(this.preferences.show_social.value ?
+				'<div id="xkit-1cp-social">' +
```
Comment (nightpool) [inline Extensions/one_click_postage.js]: template strings

File: Extensions/one_click_postage.js [hunk 5]
```diff
@@ -568,130 +455,105 @@ XKit.extensions.one_click_postage = new Object({
...
-			$("#x1cpostage_box").addClass("xkit_x1cpostage_queue_press");
-		});
-
-		$(document).on("mouseup", "#x1cpostage_queue", function() {
-			$("#x1cpostage_box").removeClass("xkit_x1cpostage_queue_press");
-		});
-
-		var m_blogselector_html = "";
+		if (this.preferences.show_reverse_ui.value) {
+			$(document).on("mouseover", "#x1cpostage_queue", function() {
```
Comment (nightpool) [inline Extensions/one_click_postage.js]: why is this inside the show_reverse_ui conditional?

File: Extensions/one_click_postage.js [hunk 6]
```diff
@@ -568,130 +455,105 @@ XKit.extensions.one_click_postage = new Object({
...
 
-		var m_remove_box_style = "";
-		if (!this.preferences.show_caption.value) {
-			m_remove_box_style = " style=\"display: none;\" ";
-		}
+		var ui_html = [
+			'<div id="x1cpostage_reblog"><i>&nbsp;</i></div>' + '<div id="x1cpostage_queue"><i>&nbsp;</i></div>' + '<div id="x1cpostage_draft"><i>&nbsp;</i></div>',
 
-		var m_clear_tags_button = "<div id=\"x1cpostage_clear_tags\">clear tags</div>";
+			(this.preferences.show_social.value ?
```
Comment (nightpool) [inline Extensions/one_click_postage.js]:
```
if (this.preferences.show_social.value) {
  ui_html.append(whatever)
}
```
is probably more readable then this repeated `conditional ? thing : ''` pattern in an array literal

File: Extensions/one_click_postage.js [hunk 7]
```diff
@@ -568,130 +455,105 @@ XKit.extensions.one_click_postage = new Object({
...
+				'<div id="x1cpostage_clear_tags">clear tags</div>'
+			: '')
+		];
 
-		$(document).on("mouseout", "#x1cpostage_queue", function() {
-			$("#x1cpostage_box").removeClass("xkit_x1cpostage_queue_press");
-			$("#x1cpostage_box").removeClass("xkit_x1cpostage_queue_hover");
-		});
+		$("body").append('<div id="x1cpostage_box"' + (this.preferences.show_reverse_ui.value ?
+			'>' + ui_html.reverse().join("") :
```
Comment (nightpool) [inline Extensions/one_click_postage.js]:
```
if (this.preferences.show_reverse_ui) {
  ui_html = ui_html.reverse();
}
```

File: Extensions/xkit_patches.js [hunk 8]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
 				});
 				return posts;
 			};
+
+			/**
+			 * @param {JQuery} $post - Post element
+			 * @return {Object} Interface Post Object or {error: true}
+			 */
+			XKit.interface.parse_post = function($post) {
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: I think renaming `interface.post` makes this pull request a lot harder to review then it needs to be, and it has the potential to break dozens of non-centralized XKit userscripts. It also leaves many confusingly-named functions in its place, like `find_post` and `get_posts`.

File: Extensions/xkit_patches.js [hunk 9]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+									resolve(JSON.parse(response.responseText));
+								} catch (e) {
+									this.onerror(response);
+								}
+							}
+						});
+					});
+				},
+
+				like: function(post) {
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: unused? probably shouldn't be part of this PR

File: Extensions/xkit_patches.js [hunk 10]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+									resolve(JSON.parse(response.responseText));
+								} catch (e) {
+									this.onerror(response);
+								}
+							}
+						});
+					});
+				},
+
+				unlike: function(post) {
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: unused, ibid

File: Extensions/xkit_patches.js [hunk 11]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+					m_error.error = true;
+					m_error.error_message = "Object not found on page.";
+					return m_error;
+				}
+
+			};
+
+			XKit.interface.post = { /* globals Promise */
+
+				fetch: function(post) {
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: Adding a new `XKit.interface.post.fetch` when we already have `XKit.interface.fetch` seems like a recipe for disaster. Why not just make that function return a promise if you don't pass it a callback?

File: Extensions/xkit_patches.js [hunk 12]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+				} else {
+					var m_error = {};
+					m_error.error = true;
+					m_error.error_message = "Object not found on page.";
+					return m_error;
+				}
+
+			};
+
+			XKit.interface.post = { /* globals Promise */
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: This should be in the eslintrc file, not here.

File: Extensions/xkit_patches.js [hunk 13]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+									} catch (e) {
+										reject(response);
+									}
+								}
+							});
+						});
+					});
+				},
+
+				delete: function(post) {
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: also unused.

File: Extensions/xkit_patches.js [hunk 14]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+							headers: {
+								"X-Tumblr-Form-Key": XKit.interface.form_key(),
+								"X-Requested-With": "XMLHttpRequest"
+							},
+							json: true, data: JSON.stringify(post),
+							onerror: function(response) {
+								reject(response);
+							},
+							onload: function(response) {
+								try {
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: this whole function can be replaced with `resolve(response.reponseText)` and `.then(JSON.parse)`, avoiding the need for a try-catch and confusing control flow.

File: Extensions/xkit_patches.js [hunk 15]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+				fetch: function(post) {
+					return new Promise(function(resolve, reject) {
+						XKit.tools.Nx_XHR({
+							method: "POST",
+							url: "https://www.tumblr.com/svc/post/fetch",
+							headers: {
+								"X-Tumblr-Form-Key": XKit.interface.form_key(),
+								"X-Requested-With": "XMLHttpRequest"
+							},
+							json: true, data: JSON.stringify(post),
```
Comment (nightpool) [inline Extensions/xkit_patches.js]:
```
json: true,
data: JSON.stringify(post),
```

File: Extensions/xkit_patches.js [hunk 16]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+
+							XKit.tools.Nx_XHR({
+								method: "POST",
+								url: "https://www.tumblr.com/svc/post/update",
+								headers: {
+									"X-Tumblr-Puppies": kitty.kitten,
+									"X-Tumblr-Form-Key": XKit.interface.form_key(),
+									"X-Requested-With": "XMLHttpRequest"
+								},
+								json: true, data: JSON.stringify(post),
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: ibid

File: Extensions/xkit_patches.js [hunk 17]
```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
...
+								onerror: function(response) {
+									reject(response);
+								},
+								onload: function(response) {
+									try {
+										resolve(JSON.parse(response.responseText));
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: ibid

File: Extensions/xkit_patches.js (hunk 8 above)
Comment (AprilSylph) [reply Extensions/xkit_patches.js]: a very valid point, but i honestly cant think of what else to name the new thing - if you have even a half-decent suggestion i can leave `interface.post` as it was

File: Extensions/xkit_patches.js (hunk 11 above)
Comment (AprilSylph) [reply Extensions/xkit_patches.js]: because that function is horrible, and part of the eventual plan is to replace it with this new one and get rid of it :P

File: Extensions/xkit_patches.js (hunk 9 above)
Comment (AprilSylph) [reply Extensions/xkit_patches.js]: all these functions are the entire point of the PR - i only touched OCP as an *example* of how nice it'll be to have them ready to put into extensions

File: Extensions/one_click_postage.js (hunk 2 above)
Comment (nightpool) [reply Extensions/one_click_postage.js]: post_type: ""

File: Extensions/xkit_patches.js (hunk 9 above)
Comment (nightpool) [reply Extensions/xkit_patches.js]: I get that, but they're still dead code. There's no guarantee they'll get used in the future, and they're very likely to break without anyone using them.

Comment (AprilSylph) [reply Extensions/xkit_patches.js]: Servant and Mass Deleter could both use these functions, and I'd rather have this merged first and be able to open PRs for both those extensions rather than choose which to do first - or I could work on them *here* and make this PR even bigger

**Comment by nightpool** [comment]
not sure I understand. are you saying that once this is merged you can't
open a new PR that adds this one function to xkit_patches and refactors
servant and mass delete to use it?

File: Extensions/xkit_patches.js (hunk 9 above)
Comment (nightpool) [reply Extensions/xkit_patches.js]: not sure I understand. whats the problem with, once this is merged, opening a new PR that adds one function to xkit_patches and refactors servent to use it?

**Comment by nightpool** [comment]
anyway, if we're pretty dead set on making this a completely separate API from the existing SVC functions, (fetch, is_following, etc), then I think grouping them as such makes sense, under something like `XKit.interface.svc_api.fetch`

File: Extensions/xkit_patches.js (hunk 9 above)
Comment (AprilSylph) [reply Extensions/xkit_patches.js]: the problem being that if I did that, I would have to either wait for that new PR to be merged before I used them in Mass Deleter, add the same functions in two separate PRs, or refactor multiple unrelated extensions in the same PR

i also dont see the point of withholding them from the codebase when they work and are unlikely to break before theyre used officially, but thats just me

**Comment by AprilSylph** [comment]
considering a few things youve said, I'd like to hear what you think of taking this PR in an entirely different direction. you're right that this PR is too big, and here's what I'd love to do instead at this point:
1) majorly simplify this PR down into just `Nx_XHR` improvements and the new Promise functions, under `XKit.interface.svc` - there's the post functions here, but we also use `/svc/indash_blog` in Timestamps and I would *love* an easier way of accessing the peepr API as well
2) a followup PR that replaces all manual svc XHRs with them, without changing the surrounding unrelated code (or given your opinion on to-be-used functions, could go here too)
3) another followup PR that replaces & removes the previous svc functions

then, any further code cleanup I feel like doing can be one PR per extension again!

tl;dr i admit i got carried away with OCP's code and i would like to start this PR again please if thats not too annoying of me

**Comment by nightpool** [comment]
@AprilSylph I don't think each PR needs to be *one* extension, but it does need to be *one* logical change. It's really important to me not to have functions before we use them, because it's hard to predict how they're going to need to be used. I think the ideal pull request size is *one* xkit_patches function and fixing *one* extension that it gets used in, and then iteratively improving on that. *two* extensions it gets used in would probably also be fine.
